### PR TITLE
Initial code with the ADU OTA logic in ESP32 sample

### PR DESCRIPTION
**Pull request by ewertons**
## Purpose

* ...

## Does this introduce a breaking change?

```
[ ] Yes
[ ] No
```

## Pull Request Type
What kind of change does this Pull Request introduce?

```
[ ] Bugfix
[ ] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Documentation content changes
[ ] Other... Please describe:
```

## How to Test
*  Get the code

```
git clone [repo-address]
cd [repo-name]
git checkout [branch-name]
npm install
```

* Test the code

```
```

## What to Check
Verify that the following are valid
* ...

### CODE REVIEW

File: demos/projects/ESPRESSIF/aduotaupdate/main/azure_iot_freertos_esp32_pnp_model.c [hunk 1]
```diff
@@ -341,28 +205,187 @@ void vHandleWritableProperties( AzureIoTHubClientPropertiesResponse_t * pxMessag
...
+            {
+                LogError( ( "[OTA] deployment cancelled, but agent is not in DIP state" ) );
+                // TODO: Fail? (send AZ_IOT_ADU_OTA_AGENT_STATE_FAILED?)
+                //       If already in FAILED state, should reset to IDLE?
+            }
+        }
+
+        xOtaServiceActionReceived = false;
+    }
+    else if ( uOtaAgentState == AZ_IOT_ADU_OTA_AGENT_STATE_DEPLOYMENT_IN_PROGRESS )
```
[the remark this answers is not in the document]
Comment (danewalton) [reply demos/projects/ESPRESSIF/aduotaupdate/main/azure_iot_freertos_esp32_pnp_model.c:368]: or is it not supposed to?

File: libs/azure-iot-middleware-freertos/libraries/azure-sdk-for-c/sdk/src/azure/iot/az_iot_adu_ota.c [hunk 2]
```diff
@@ -0,0 +1,478 @@
...
+                required_size = jr->token.size + NULL_TERM_CHAR_SIZE;
+
+                _az_RETURN_IF_NOT_ENOUGH_SIZE(buffer, required_size);
+
+                ota_state->_internal.file_urls[ota_state->_internal.file_urls_count]._internal.id = split_az_span(buffer, required_size, &buffer);
+
+                _az_RETURN_IF_FAILED(az_json_token_get_string(
+                    &jr->token, (char*)az_span_ptr(ota_state->_internal.file_urls[ota_state->_internal.file_urls_count]._internal.id),
+                    az_span_size(ota_state->_internal.file_urls[ota_state->_internal.file_urls_count]._internal.id), NULL));
+
```
[the remark this answers is not in the document]
Comment (danewalton) [reply libs/azure-iot-middleware-freertos/libraries/azure-sdk-for-c/sdk/src/azure/iot/az_iot_adu_ota.c:325]: yayyyyy working with C :)

File: libs/azure-iot-middleware-freertos/libraries/azure-sdk-for-c/sdk/inc/azure/iot/az_iot_adu_ota.h [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+
+AZ_NODISCARD az_result az_iot_adu_ota_get_workflow(az_iot_adu_ota_state* ota_state, az_span* id, int32_t* action, az_span* retry_timestamp);
+
+AZ_NODISCARD az_result az_iot_adu_ota_get_update_manifest(az_iot_adu_ota_state* ota_state, az_span* update_manifest);
+
+AZ_NODISCARD az_result az_iot_adu_ota_get_update_manifest_signature(az_iot_adu_ota_state* ota_state, az_span* update_manifest_signature);
+
+AZ_NODISCARD az_result az_iot_adu_ota_get_file_url_count(az_iot_adu_ota_state* ota_state, uint32_t* file_url_count);
+
+AZ_NODISCARD az_result az_iot_adu_ota_get_file_url(az_iot_adu_ota_state* ota_state, uint32_t index, az_span* file_id, az_span* file_url);
```
Comment (danewalton) [inline libs/azure-iot-middleware-freertos/libraries/azure-sdk-for-c/sdk/inc/azure/iot/az_iot_adu_ota.h:109]: nit of nits: I would flip these (order).

**Review by danewalton — APPROVED** [review]
Looks overall pretty good. Thanks for all the work @ewertons !

File: libs/azure-iot-middleware-freertos/libraries/azure-sdk-for-c/sdk/inc/azure/iot/az_iot_adu_ota.h [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+/* Maximum Number of Files Handled by this ADU Agent */
+#define AZ_IOT_ADU_OTA_FILE_URL_MAX_COUNT 10
+
+typedef struct
+{
+    az_span manufacturer;
+    az_span model;
+    az_span last_installed_update_id;
+    az_span adu_version;
+    az_span do_version;
```
[the remark this answers is not in the document]
Comment (ewertons) [reply libs/azure-iot-middleware-freertos/libraries/azure-sdk-for-c/sdk/inc/azure/iot/az_iot_adu_ota.h:48]: We will review all the names for the final api (after PoC).
For now, I added a link to the ADU documentation that mentions what doVer means.

File: libs/azure-iot-middleware-freertos/libraries/azure-sdk-for-c/sdk/inc/azure/iot/az_iot_adu_ota.h [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+#include <stdbool.h>
+#include <stdint.h>
+
+#include <azure/core/_az_cfg_prefix.h>
+
+/* ADU PnP Component Name */
+#define AZ_IOT_ADU_PROPERTIES_COMPONENT_NAME                 "deviceUpdate"
+
+/* ADU Service Actions */
+#define AZ_IOT_ADU_OTA_SERVICE_ACTION_APPLY_DEPLOYMENT       3
```
[the remark this answers is not in the document]
Comment (ewertons) [reply libs/azure-iot-middleware-freertos/libraries/azure-sdk-for-c/sdk/inc/azure/iot/az_iot_adu_ota.h:30]: Those are values valid for agents before version 0.8.0 (pre-PPR).
Now the actions are just "ApplyDeployment" and "Cancel".

[the remark this answers is not in the document]
Comment (ewertons) [reply libs/azure-iot-middleware-freertos/libraries/azure-sdk-for-c/sdk/inc/azure/iot/az_iot_adu_ota.h:30]: So, no, we should not include the older actions.

File: demos/sample_azure_iot/sample_azure_iot.c [hunk 6]
```diff
@@ -362,6 +362,13 @@ static void prvAzureDemoTask( void * pvParameters )
         xHubOptions.pucModuleID = ( const uint8_t * ) democonfigMODULE_ID;
         xHubOptions.ulModuleIDLength = sizeof( democonfigMODULE_ID ) - 1;
 
+        // ADU OTA
```
[the remark this answers is not in the document]
Comment (ewertons) [reply demos/sample_azure_iot/sample_azure_iot.c]: Oops. That would have been a good idea. But I missed that, sorry.

File: libs/azure-iot-middleware-freertos/libraries/azure-sdk-for-c/sdk/inc/azure/iot/az_iot_adu_ota.h [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+    az_span adu_version;
+    az_span do_version;
+} az_iot_adu_ota_device_information;
+
+typedef struct
+{
+    struct
+    {
+        int32_t action;
+        az_span id;
```
[the remark this answers is not in the document]
Comment (ewertons) [reply libs/azure-iot-middleware-freertos/libraries/azure-sdk-for-c/sdk/inc/azure/iot/az_iot_adu_ota.h]: We will add documentation later. This is the PoC of the PoC. 
This ID field is in the workflow structure.
If you refer to the ADU documentation, it's straightforward getting the meaning for now. 
https://docs.microsoft.com/en-us/azure/iot-hub-device-update/device-update-plug-and-play

File: libs/azure-iot-middleware-freertos/libraries/azure-sdk-for-c/sdk/inc/azure/iot/az_iot_adu_ota.h [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+        az_span retry_timestamp;
+    } _internal;
+} az_iot_adu_ota_workflow;
+
+typedef struct
+{
+    struct
+    {
+        az_span id;
```
[the remark this answers is not in the document]
Comment (ewertons) [reply libs/azure-iot-middleware-freertos/libraries/azure-sdk-for-c/sdk/inc/azure/iot/az_iot_adu_ota.h]: Same here, this is the poc of the poc. 
For details for now, here is the doc: https://docs.microsoft.com/en-us/azure/iot-hub-device-update/device-update-plug-and-play
We will add the documentation soon.

File: demos/projects/ESPRESSIF/aduotaupdate/main/azure_iot_freertos_esp32_pnp_model.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (ewertons) [reply demos/projects/ESPRESSIF/aduotaupdate/main/azure_iot_freertos_esp32_pnp_model.c:391]: We could, but this is a hack state machine just to simulate the transitions of the OTA workflow.
It's ok to have just one operation per invocation of ulSampleCreateReportedPropertiesUpdate

File: libs/azure-iot-middleware-freertos/libraries/azure-sdk-for-c/sdk/src/azure/iot/az_iot_adu_ota.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (ewertons) [reply libs/azure-iot-middleware-freertos/libraries/azure-sdk-for-c/sdk/src/azure/iot/az_iot_adu_ota.c]: Fixed. Yes, that was just a rough first implementation. But I improved that already to parse properties in any order.

File: demos/projects/ESPRESSIF/aduotaupdate/main/azure_iot_freertos_esp32_pnp_model.c [hunk 9]
```diff
@@ -341,28 +205,187 @@ void vHandleWritableProperties( AzureIoTHubClientPropertiesResponse_t * pxMessag
...
+                    return;
+                }
+                else
+                {
+                    * pulWritablePropertyResponseBufferLength = az_span_size( xWritablePropertyResponse ) ;
+                    xOtaServiceActionReceived = true;
+                }
+            }
+
+
```
[the remark this answers is not in the document]
Comment (ewertons) [reply demos/projects/ESPRESSIF/aduotaupdate/main/azure_iot_freertos_esp32_pnp_model.c]: Fixed.

File: demos/projects/ESPRESSIF/aduotaupdate/main/azure_iot_freertos_esp32_pnp_model.c [hunk 10]
```diff
@@ -341,28 +205,187 @@ void vHandleWritableProperties( AzureIoTHubClientPropertiesResponse_t * pxMessag
...
+                                                                                  &pucComponentName, &ulComponentNameLength ) ) == eAzureIoTSuccess )
+    {
+        LogInfo( ( "Properties component name: %.*s", ulComponentNameLength, pucComponentName ) );
+
+        if ( az_iot_adu_ota_is_component_device_update( az_span_create( ( uint8_t* ) pucComponentName, ulComponentNameLength ) ) )
+        {
+            az_json_reader* jr = &xJsonReader._internal.xCoreReader;
+            LogInfo( ( "Device update received: %.*s\r\n", ulComponentNameLength, pucComponentName ) );
+
+            if ( az_result_failed( az_iot_adu_ota_parse_service_properties( &xOtaState, jr, AZ_SPAN_FROM_BUFFER( uOtaBuffer ), &uOtaBufferRemainder ) ) )
```
[the remark this answers is not in the document]
Comment (ewertons) [reply demos/projects/ESPRESSIF/aduotaupdate/main/azure_iot_freertos_esp32_pnp_model.c]: Intentional at this point.

File: libs/azure-iot-middleware-freertos/libraries/azure-sdk-for-c/sdk/inc/azure/iot/az_iot_adu_ota.h (hunk 3 above)
Comment (ewertons) [reply libs/azure-iot-middleware-freertos/libraries/azure-sdk-for-c/sdk/inc/azure/iot/az_iot_adu_ota.h]: These "get" functions are now gone.

File: libs/azure-iot-middleware-freertos/libraries/azure-sdk-for-c/sdk/src/azure/iot/az_iot_adu_ota.c [hunk 11]
```diff
@@ -0,0 +1,478 @@
...
+    return AZ_OK;
+}
+
+AZ_NODISCARD az_result az_iot_adu_ota_get_properties_payload(
+    az_iot_adu_ota_state* ota_state, int32_t agent_state, az_span payload, az_span* out_payload)
+{
+    _az_PRECONDITION_NOT_NULL(ota_state);
+    _az_PRECONDITION_VALID_SPAN(payload, 1, false);
+    _az_PRECONDITION_NOT_NULL(out_payload);
```
[the remark this answers is not in the document]
Comment (ewertons) [reply libs/azure-iot-middleware-freertos/libraries/azure-sdk-for-c/sdk/src/azure/iot/az_iot_adu_ota.c:141]: ota state is now gone. But your comment will still apply to the other structures we have (later in the next PRs.)

File: libs/azure-iot-middleware-freertos/libraries/azure-sdk-for-c/sdk/inc/azure/iot/az_iot_adu_ota.h [hunk 12]
```diff
@@ -0,0 +1,120 @@
...
+        az_span url;
+    } _internal;
+} az_iot_adu_ota_file_url;
+
+typedef struct
+{
+  struct
+  {
+    /* Device Qualifications */
+    az_iot_hub_client const* client;
```
[the remark this answers is not in the document]
Comment (ewertons) [reply libs/azure-iot-middleware-freertos/libraries/azure-sdk-for-c/sdk/inc/azure/iot/az_iot_adu_ota.h]: That structure is now removed. The code is not keeping an OTA "state" anymore.

File: libs/azure-iot-middleware-freertos/libraries/azure-sdk-for-c/sdk/src/azure/iot/az_iot_adu_ota.c [hunk 13]
```diff
@@ -0,0 +1,478 @@
...
+
+
+/* --- az_core extensions --- */
+static az_span split_az_span(az_span span, int32_t size, az_span* remainder)
+{
+  az_span result = az_span_slice(span, 0, size);
+
+  if (remainder != NULL && !az_span_is_content_equal(AZ_SPAN_EMPTY, result))
+  {
+    *remainder = az_span_slice(span, size, az_span_size(span));
```
[the remark this answers is not in the document]
Comment (ewertons) [reply libs/azure-iot-middleware-freertos/libraries/azure-sdk-for-c/sdk/src/azure/iot/az_iot_adu_ota.c:507]: That is a good idea. I'll do that.

File: libs/azure-iot-middleware-freertos/libraries/azure-sdk-for-c/sdk/src/azure/iot/az_iot_adu_ota.c [hunk 14]
```diff
@@ -0,0 +1,478 @@
...
+}
+
+AZ_NODISCARD az_result az_iot_adu_ota_get_service_properties_response(az_iot_adu_ota_state* ota_state, int32_t version, int32_t status, az_span payload, az_span* out_payload)
+{
+    _az_PRECONDITION_NOT_NULL(ota_state);
+    _az_PRECONDITION_VALID_SPAN(payload, 1, false);
+    _az_PRECONDITION_NOT_NULL(out_payload);
+
+    az_json_writer jw;
+
```
[the remark this answers is not in the document]
Comment (ewertons) [reply libs/azure-iot-middleware-freertos/libraries/azure-sdk-for-c/sdk/src/azure/iot/az_iot_adu_ota.c:457]: Taken. Thx

File: demos/projects/ESPRESSIF/aduotaupdate/main/azure_iot_freertos_esp32_pnp_model.c [hunk 15]
```diff
@@ -341,28 +205,187 @@ void vHandleWritableProperties( AzureIoTHubClientPropertiesResponse_t * pxMessag
...
+        LogError( ( "AzureIoTJSONReader_Init failed: result 0x%08x", xAzIoTResult ) );
+        * pulWritablePropertyResponseBufferLength = 0;
+        return;
+    }
+
+    xAzIoTResult = AzureIoTHubClientProperties_GetPropertiesVersion( &xAzureIoTHubClient, &xJsonReader, pxMessage->xMessageType, &ulPropertyVersion );
+    if ( xAzIoTResult != eAzureIoTSuccess )
+    {
+        LogError( ( "AzureIoTHubClientProperties_GetPropertiesVersion failed: result 0x%08x", xAzIoTResult ) );
+        * pulWritablePropertyResponseBufferLength = 0;
```
[the remark this answers is not in the document]
Comment (ewertons) [reply demos/projects/ESPRESSIF/aduotaupdate/main/azure_iot_freertos_esp32_pnp_model.c:253]: Done.

File: demos/projects/ESPRESSIF/aduotaupdate/main/azure_iot_freertos_esp32_pnp_model.c (hunk 10 above)
[the remark this answers is not in the document]
Comment (ewertons) [reply demos/projects/ESPRESSIF/aduotaupdate/main/azure_iot_freertos_esp32_pnp_model.c]: Yeah. You know, PoC stuff. But I removed it.

File: demos/projects/ESPRESSIF/aduotaupdate/main/azure_iot_freertos_esp32_pnp_model.c [hunk 16]
```diff
@@ -78,168 +79,31 @@ static const char *TAG = "sample_azureiotkit";
 #define sampleazureiotPROPERTY_TELEMETRY_FREQUENCY ( "telemetryFrequencySecs" )
 
 static int lTelemetryFrequencySecs = 2;
+
 /*-----------------------------------------------------------*/
+// ADU OTA
+#define OTA_BUFFER_SIZE 5000
```
[the remark this answers is not in the document]
Comment (ewertons) [reply demos/projects/ESPRESSIF/aduotaupdate/main/azure_iot_freertos_esp32_pnp_model.c:85]: I am not sure these names are the best as well, but I'll take any suggestions for now. Names will be reviewed afterwards.

File: demos/projects/ESPRESSIF/aduotaupdate/main/azure_iot_freertos_esp32_pnp_model.c [hunk 17]
```diff
@@ -78,168 +79,31 @@ static const char *TAG = "sample_azureiotkit";
...
+static uint8_t uOtaBuffer[OTA_BUFFER_SIZE];
```
[the remark this answers is not in the document]
Comment (ewertons) [reply demos/projects/ESPRESSIF/aduotaupdate/main/azure_iot_freertos_esp32_pnp_model.c:86]: Same as above.
Names will be reviewed afterwards.

**Pull request merged**